File: ruby/proverb/proverb.rb
class Proverb
  def initialize(*subjects, qualifier: nil)
    @proverb = subjects
               .each_cons(2)
               .map { |(want, lost)| "For want of a #{want} the #{lost} was lost." }
               .push("And all for the want of a #{qualifier} #{subjects.first}.".squeeze(' '))
               .join("\n")
  end

  def to_s
    @proverb
  end
end

File: clojure/proverb/src/proverb.clj
(ns proverb (:require [clojure.string :as str]))

(defn recite [pieces]
  (if
    (empty? pieces) ""
    (->> pieces
         (partition 2 1)
         (map #(str "For want of a " (first %) " the " (second %) " was lost."))
         (#(concat % [(str "And all for the want of a " (first pieces) ".")]))
         (str/join "\n"))))
